compositor_pipeline/src/queue.rs

```diff
@@ -0,0 +1,94 @@
+mod internal_queue;
+
+use std::{
+    sync::{Arc, Mutex},
+    thread::{self},
+    time::{Duration, Instant},
+};
+
+use compositor_common::{
+    frame::{Framerate, FramesBatch, InputID},
+    Frame,
+};
+use crossbeam_channel::{tick, unbounded, Receiver, Sender};
+
+use self::internal_queue::{InternalQueue, QueueError};
+
+pub struct Queue {
+    internal_queue: Arc<Mutex<InternalQueue>>,
+    check_queue_channel: (Sender<()>, Receiver<()>),
+    output_framerate: Framerate,
+}
+
+impl Queue {
+    #[allow(dead_code)]
+    pub fn new(output_framerate: Framerate) -> Self {
+        Queue {
+            internal_queue: Arc::new(Mutex::new(InternalQueue::new(output_framerate))),
+            check_queue_channel: unbounded(),
+            output_framerate,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn add_input(&self, input_id: InputID) {
+        let mut internal_queue = self.internal_queue.lock().unwrap();
+        internal_queue.add_input(input_id);
+    }
+
+    #[allow(dead_code)]
+    pub fn remove_input(&self, input_id: InputID) {
+        let mut internal_queue = self.internal_queue.lock().unwrap();
+        // TODO: gracefully remove input - wait until last enqueued frame PTS is smaller than output PTS
+        internal_queue.remove_input(input_id);
+    }
+
+    #[allow(dead_code)]
+    pub fn start(&self, sender: Sender<FramesBatch>) {
+        let (check_queue_sender, check_queue_receiver) = self.check_queue_channel.clone();
+        let internal_queue = self.internal_queue.clone();
+        let tick_duration = self.output_framerate.get_interval_duration();
+
+        thread::spawn(move || {
+            // Wait for first frame
+            check_queue_receiver.recv().unwrap();
+            Self::start_ticker(tick_duration, check_queue_sender);
+
+            let start = Instant::now();
+            loop {
+                let mut internal_queue = internal_queue.lock().unwrap();
+                let buffer_pts = internal_queue.get_next_output_buffer_pts();
+
+                if start.elapsed() > buffer_pts || internal_queue.check_all_inputs_ready(buffer_pts)
```

Comment:
I don't see anything related to the additional buffer we want to keep. shouldn't this be 
start.elapsed() + fixed_buffer_duration > buffer_pts

or I guess this could be taken into account when enqueueing

Reply:
Fixed